crates/sui-snapshot/src/uploader.rs

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::writer::StateSnapshotWriterV1;
+use anyhow::Result;
+use bytes::Bytes;
+use object_store::DynObjectStore;
+use oneshot::channel;
+use prometheus::{register_int_gauge_with_registry, IntGauge, Registry};
+use std::num::NonZeroUsize;
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::time::Duration;
+use sui_core::authority::authority_store_tables::AuthorityPerpetualTables;
+use sui_core::db_checkpoint_handler::{STATE_SNAPSHOT_COMPLETED_MARKER, SUCCESS_MARKER};
+use sui_protocol_config::ProtocolConfig;
+use sui_storage::object_store::util::{
+    find_all_dirs_with_epoch_prefix, find_missing_epochs_dirs, path_to_filesystem, put,
+};
+use sui_storage::object_store::{ObjectStoreConfig, ObjectStoreType};
+use sui_storage::FileCompression;
+use tokio::sync::oneshot;
+use tokio::sync::oneshot::Sender;
+use tracing::{error, info};
+
+pub struct StateSnapshotUploaderMetrics {
+    pub first_missing_state_snapshot_epoch: IntGauge,
+}
+
+impl StateSnapshotUploaderMetrics {
+    pub fn new(registry: &Registry) -> Arc<Self> {
+        let this = Self {
+            first_missing_state_snapshot_epoch: register_int_gauge_with_registry!(
+                "first_missing_state_snapshot_epoch",
+                "First epoch for which we have no state snapshot in remote store",
+                registry
+            )
+            .unwrap(),
+        };
+        Arc::new(this)
+    }
+}
+
+pub struct StateSnapshotUploader {
+    /// Directory path on local disk where db checkpoints are stored
+    local_input_path: PathBuf,
+    /// Directory on local disk where db checkpoints are stored
```

Comment:
Aren't these `DynObjectStore` objects, not paths? Similar with the other store fields below